File: examples/langchain_typescript/retrieval_agent.ts
/**
 * LangChain TypeScript retrieval agent with vector search
 */

import { AgentExecutor } from "langchain/agents";
import { ChatOpenAI, OpenAIEmbeddings } from "@langchain/openai";
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { MemoryVectorStore } from "langchain/vectorstores/memory";

// Define retrieval tools
const searchDocumentation = tool({
  name: "search_documentation",
  description: "Search the product documentation using semantic search",
  schema: z.object({
    query: z.string().describe("Search query"),
  }),
  func: async ({ query }) => {
    // Mock implementation
    return `Documentation results for '${query}': Found 5 relevant articles about authentication`;
  },
});

const searchKnowledgeBase = tool({
  name: "search_knowledge_base",
  description: "Search the internal knowledge base using vector search",
  schema: z.object({
    query: z.string().describe("Search query"),
  }),
  func: async ({ query }) => {
    // Mock implementation
    return `Knowledge base results: Best practices for ${query}`;
  },
});

const retrieveSimilarCases = tool({
  name: "retrieve_similar_cases",
  description: "Retrieve similar support cases based on description",
  schema: z.object({
    description: z.string().describe("Case description"),
  }),
  func: async ({ description }) => {
    // Mock implementation
    return `Found 3 similar cases related to: ${description}`;
  },
});

const getContext = tool({
  name: "get_context",
  description: "Get relevant context for a specific topic from vectorstore",
  schema: z.object({
    topic: z.string().describe("Topic to get context for"),
  }),
  func: async ({ topic }) => {
    // Mock implementation
    return `Context for ${topic}: Key concepts and examples`;
  },
});

// Initialize embeddings
const embeddings = new OpenAIEmbeddings();

// Initialize the language model
const llm = new ChatOpenAI({
  modelName: "gpt-4",
  temperature: 0.7,
});

// Create the tools array
const tools = [
  searchDocumentation,
  searchKnowledgeBase,
  retrieveSimilarCases,
  getContext,
];

// Create the retrieval agent executor
const retrievalAgent = new AgentExecutor({
  tools: tools,
  agent: llm,
  verbose: true,
  maxIterations: 8,
});

// Example usage
async function main() {
  const result = await retrievalAgent.invoke({
    input: "Find documentation about user authentication and any similar support cases"
  });
  console.log(result);
}

if (require.main === module) {
  main();
}

export { retrievalAgent };

File: examples/langchain_python/retrieval_agent.py
"""LangChain retrieval agent with vector search capabilities."""

from langchain.agents import initialize_agent, AgentType
from langchain.tools import tool
from langchain_openai import ChatOpenAI, OpenAIEmbeddings
from langchain_community.vectorstores import FAISS

@tool
def search_documentation(query: str) -> str:
    """Search the product documentation using semantic search."""
    # Mock implementation
    return f"Documentation results for '{query}': Found 5 relevant articles about authentication"

@tool
def search_knowledge_base(query: str) -> str:
    """Search the internal knowledge base using vector search."""
    # Mock implementation
    return f"Knowledge base results: Best practices for {query}"

@tool
def retrieve_similar_cases(description: str) -> str:
    """Retrieve similar support cases based on description."""
    # Mock implementation
    return f"Found 3 similar cases related to: {description}"

@tool
def get_context(topic: str) -> str:
    """Get relevant context for a specific topic from vectorstore."""
    # Mock implementation
    return f"Context for {topic}: Key concepts and examples"

# Initialize embeddings
embeddings = OpenAIEmbeddings()

# Initialize the language model
llm = ChatOpenAI(
    model="gpt-4",
    temperature=0.7
)

# Create the tools list
tools = [
    search_documentation,
    search_knowledge_base,
    retrieve_similar_cases,
    get_context
]

# Initialize the retrieval agent
retrieval_agent = initialize_agent(
    tools=tools,
    llm=llm,
    agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
    agent_name="Knowledge Retrieval Agent",
    verbose=True,
    handle_parsing_errors=True
)

# Example usage
if __name__ == "__main__":
    response = retrieval_agent.run(
        "Find documentation about user authentication and any similar support cases"
    )
    print(response)

